### PR TITLE
Groovy extension is activated even though vertx-lang-groovy is not in the classpath

### PR DESCRIPTION
**Pull request by cazacugmihai**
Fixed [this](https://github.com/vert-x3/vertx-lang-groovy/issues/48) issue.

For tests, I've just run `mvn test` but I don't know if this is enough.

### CODE REVIEW

**Comment by cazacugmihai** [comment]
Can someone take a look over this PR, please? My project keeps crashing  because of this. Thanks!

File: src/main/resources/vertx-groovy/template/gen-module-factory.templ [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import java.util.Collections;\n
+import java.util.List;\n
+import java.util.Properties;\n
+\n
+public class @{type.simpleName}_GroovyPropertiesModuleFactory extends PropertiesModuleFactory {\n
+\n
+  @Override\n
+  public ExtensionModule newModule(Properties properties, ClassLoader classLoader) {\n
+    try {\n
+      Class.forName("io.vertx.lang.groovy.GroovyVerticleFactory");\n
```
Comment (vietj) [inline src/main/resources/vertx-groovy/template/gen-module-factory.templ]: rather use the provided classLoader I think in the params

**Comment by vietj** [comment]
I'm also wondering how it works in fatjars, can you check fatjar creation ?

**Comment by vietj** [comment]
does it fix your case btw ?

File: src/main/resources/vertx-groovy/template/gen-module-factory.templ (hunk 1 above)
Comment (cazacugmihai) [inline src/main/resources/vertx-groovy/template/gen-module-factory.templ]: Done. Please re-check it.

**Comment by cazacugmihai** [comment]
I don't know how to check this fix in relation to [vertx-sql-common](https://github.com/vert-x3/vertx-sql-common) module. 

All I have to do is to compile `vertx-sql-common` using this version of `vertx-lang-groovy`? I saw that there is a [relation](https://github.com/vert-x3/vertx-sql-common/blob/master/pom.xml#L41) between this two modules.

**Comment by vietj** [comment]
yes, you install the a snapshot of this module and then install a snapshot of sql-common that should have the generated module factory

**Comment by cazacugmihai** [comment]
It seems that there is also a compile-check issue (when using groovy.transform.CompileStatic):

> Error:(288, 3) Groovyc: [Static type checking] - Cannot call io.vertx.ext.sql.SQLConnection#batchWithParams(java.lang.String, java.util.List <java.util.List>, io.vertx.core.Handler <io.vertx.core.AsyncResult>) with arguments [java.lang.String, java.util.List <io.vertx.core.json.JsonArray>, io.vertx.core.Handler <io.vertx.core.AsyncResult>]

**Comment by cazacugmihai** [comment]
"Does it fix your case btw?"

Yes, it's all OK right now, excepting that sometimes it throws the mentioned exception when using `@CompileStatic`.

**Comment by vietj** [comment]
it's not clear if it fixes the issue or not, can you clarify ?

**Comment by cazacugmihai** [comment]
It fixes the [issue](https://github.com/vert-x3/vertx-lang-groovy/issues/48) that has started these discussions: 

> java.lang.ClassCastException: io.vertx.core.json.JsonArray cannot be cast to java.util.List
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193) ~[?:1.8.0_121]
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1374) ~[?:1.8.0_121]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481) ~[?:1.8.0_121]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471) ~[?:1.8.0_121]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708) ~[?:1.8.0_121]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:1.8.0_121]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499) ~[?:1.8.0_121]
	at io.vertx.groovy.ext.sql.SQLConnection_GroovyExtension.batchWithParams(SQLConnection_GroovyExtension.java:73) ~[vertx-sql-common-3.4.2-SNAPSHOT.jar:?]

But now, once in awhile, it throws a new exception:

> Cannot call io.vertx.ext.sql.SQLConnection#batchWithParams(java.lang.String, java.util.List <java.util.List>, io.vertx.core.Handler <io.vertx.core.AsyncResult>) with arguments [java.lang.String, java.util.List <io.vertx.core.json.JsonArray>, io.vertx.core.Handler <io.vertx.core.AsyncResult>]

**Comment by cazacugmihai** [comment]
And [this](https://github.com/vert-x3/vertx-sql-common/issues/44#issuecomment-293535648) script it runs now without issues.

**Comment by vietj** [comment]
great

**Comment by cazacugmihai** [comment]
Is there a resolution on this?

About fatjars: I don't see any issue with them (I've tested on my app).

**Comment by vietj** [comment]
@cazacugmihai thanks for the contribution, it will be merged soon and shall be in 3.4.3

**Comment by sfitts** [comment]
@cazacugmihai thanks for this.  The associated issue has us blocked from upgrading so I tested out this fix and it works great.  Can't wait for 3.4.2

**Comment by vietj** [comment]
I reworked the PR to add an integration test and also generate only one module factory per vertx gen module. It has been merged in master.

Would you mind to check one more time with the changes ?

**Comment by cazacugmihai** [comment]
The [reproducer](https://github.com/vert-x3/vertx-sql-common/issues/44#issuecomment-293535648) script runs fine but I have a compiler error using this version within my projects:

```
startup failed:
General error during class generation: io.vertx.groovy.ext.auth.VertxPropertiesModuleFactory cannot be cast to org.codehaus.groovy.runtime.m12n.PropertiesModuleFactory

java.lang.ClassCastException: io.vertx.groovy.ext.auth.VertxPropertiesModuleFactory cannot be cast to org.codehaus.groovy.runtime.m12n.PropertiesModuleFactory
	at org.codehaus.groovy.runtime.m12n.StandardPropertiesModuleFactory.newModule(StandardPropertiesModuleFactory.java:41)
	at org.codehaus.groovy.runtime.m12n.ExtensionModuleScanner.scanExtensionModuleFromProperties(ExtensionModuleScanner.java:79)
	at org.codehaus.groovy.runtime.m12n.ExtensionModuleScanner.scanExtensionModuleFromMetaInf(ExtensionModuleScanner.java:74)
	at org.codehaus.groovy.runtime.m12n.ExtensionModuleScanner.scanClasspathModules(ExtensionModuleScanner.java:56)
	at org.codehaus.groovy.transform.stc.StaticTypeCheckingSupport$ExtensionMethodCache.getExtensionMethods(StaticTypeCheckingSupport.java:1987)
	at org.codehaus.groovy.transform.stc.StaticTypeCheckingSupport.findDGMMethodsForClassNode(StaticTypeCheckingSupport.java:208)
	at org.codehaus.groovy.transform.stc.StaticTypeCheckingSupport.findDGMMethodsForClassNode(StaticTypeCheckingSupport.java:194)
	at org.codehaus.groovy.transform.stc.StaticTypeCheckingSupport.findDGMMethodsByNameAndArguments(StaticTypeCheckingSupport.java:923)
	at org.codehaus.groovy.transform.stc.StaticTypeCheckingSupport.findDGMMethodsByNameAndArguments(StaticTypeCheckingSupport.java:910)
	at org.codehaus.groovy.transform.sc.transformers.BooleanExpressionTransformer$OptimizingBooleanExpression.visit(BooleanExpressionTransformer.java:155)
	at org.codehaus.groovy.classgen.asm.BinaryExpressionHelper.evaluateNormalTernary(BinaryExpressionHelper.java:843)
	at org.codehaus.groovy.classgen.asm.BinaryExpressionHelper.evaluateTernary(BinaryExpressionHelper.java:868)
	at org.codehaus.groovy.classgen.AsmClassGenerator.visitTernaryExpression(AsmClassGenerator.java:644)
	at org.codehaus.groovy.ast.expr.TernaryExpression.visit(TernaryExpression.java:45)
	at org.codehaus.groovy.classgen.asm.BinaryExpressionHelper.evaluateNormalTernary(BinaryExpressionHelper.java:855)
	at org.codehaus.groovy.classgen.asm.BinaryExpressionHelper.evaluateTernary(BinaryExpressionHelper.java:868)
	at org.codehaus.groovy.classgen.AsmClassGenerator.visitTernaryExpression(AsmClassGenerator.java:644)
	at org.codehaus.groovy.ast.expr.TernaryExpression.visit(TernaryExpression.java:45)
	at org.codehaus.groovy.classgen.asm.BinaryExpressionHelper.evaluateNormalTernary(BinaryExpressionHelper.java:855)
	at org.codehaus.groovy.classgen.asm.BinaryExpressionHelper.evaluateTernary(BinaryExpressionHelper.java:868)
	at org.codehaus.groovy.classgen.AsmClassGenerator.visitTernaryExpression(AsmClassGenerator.java:644)
	at org.codehaus.groovy.ast.expr.TernaryExpression.visit(TernaryExpression.java:45)
	at org.codehaus.groovy.classgen.AsmClassGenerator.visitCastExpression(AsmClassGenerator.java:743)
	at org.codehaus.groovy.ast.expr.CastExpression.visit(CastExpression.java:87)
	at org.codehaus.groovy.classgen.asm.StatementWriter.writeReturn(StatementWriter.java:590)
	at org.codehaus.groovy.classgen.AsmClassGenerator.visitReturnStatement(AsmClassGenerator.java:632)
	at org.codehaus.groovy.ast.stmt.ReturnStatement.visit(ReturnStatement.java:49)
	at org.codehaus.groovy.classgen.asm.StatementWriter.writeBlockStatement(StatementWriter.java:85)
	at org.codehaus.groovy.classgen.asm.sc.StaticTypesStatementWriter.writeBlockStatement(StaticTypesStatementWriter.java:65)
	at org.codehaus.groovy.classgen.AsmClassGenerator.visitBlockStatement(AsmClassGenerator.java:582)
	at org.codehaus.groovy.ast.stmt.BlockStatement.visit(BlockStatement.java:71)
	at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitClassCodeContainer(ClassCodeVisitorSupport.java:110)
	at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitConstructorOrMethod(ClassCodeVisitorSupport.java:121)
	at org.codehaus.groovy.classgen.AsmClassGenerator.visitStdMethod(AsmClassGenerator.java:442)
	at org.codehaus.groovy.classgen.AsmClassGenerator.visitConstructorOrMethod(AsmClassGenerator.java:395)
	at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitMethod(ClassCodeVisitorSupport.java:132)
	at org.codehaus.groovy.classgen.AsmClassGenerator.visitMethod(AsmClassGenerator.java:523)
	at org.codehaus.groovy.ast.ClassNode.visitContents(ClassNode.java:1085)
	at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitClass(ClassCodeVisitorSupport.java:54)
	at org.codehaus.groovy.classgen.AsmClassGenerator.visitClass(AsmClassGenerator.java:233)
	at org.codehaus.groovy.control.CompilationUnit$18.call(CompilationUnit.java:830)
	at org.codehaus.groovy.control.CompilationUnit.applyToPrimaryClassNodes(CompilationUnit.java:1070)
	at org.codehaus.groovy.control.CompilationUnit.doPhaseOperation(CompilationUnit.java:613)
	at org.codehaus.groovy.control.CompilationUnit.processPhaseOperations(CompilationUnit.java:591)
	at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:568)
	at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:547)
	at org.gradle.api.internal.tasks.compile.ApiGroovyCompiler.execute(ApiGroovyCompiler.java:174)
	at org.gradle.api.internal.tasks.compile.ApiGroovyCompiler.execute(ApiGroovyCompiler.java:56)
	at org.gradle.api.internal.tasks.compile.daemon.AbstractDaemonCompiler$CompilerWorkerAdapter.execute(AbstractDaemonCompiler.java:73)
	at org.gradle.api.internal.tasks.compile.daemon.AbstractDaemonCompiler$CompilerWorkerAdapter.execute(AbstractDaemonCompiler.java:64)
	at org.gradle.workers.internal.WorkerDaemonServer.execute(WorkerDaemonServer.java:29)
	at org.gradle.api.internal.tasks.compile.daemon.AbstractDaemonCompiler$CompilerDaemonServer.execute(AbstractDaemonCompiler.java:91)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.process.internal.worker.request.WorkerAction.run(WorkerAction.java:88)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:147)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:129)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:404)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:46)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

Tested with:
- groovy version: `2.4.7`, `2.4.8`, `2.4.9`, `2.4.10`, `2.5.0-alpha-1`
- vertx version: `3.4.2-SNAPSHOT`
- vertx-jdbc-client version: `3.4.2-SNAPSHOT`

**Comment by cazacugmihai** [comment]
It seems to be a groovy's issue.

**Comment by vietj** [comment]
I haven't changed much @cazacugmihai just renamed the class and generate less classes (one per module)

**Comment by cazacugmihai** [comment]
Yes, I saw. There is something strange with groovy.  I'm trying to find the cause. Maybe with some help from @blackdrag.

**Comment by sfitts** [comment]
@cazacugmihai learn anything more about this?

I'm also wondering about this approach and fat jars (specifically as constructed by the gradle shadowJar plugin).  How does merging of module factories work?  The shadowJar plug-in doesn't seem to pay any attention to the `moduleFactory` key in the properties file and even if it did how would you resolve multiple, competing factories (for example we use RxGroovy, which has a module factory as well).

I feel like I'm missing something obvious here, but I don't see how to resolve potential factory collisions.

**Pull request merged**